kernel/src/fs/utils/id_bitmap.rs

```diff
@@ -0,0 +1,184 @@
+            first_available_id: 0,
+            len,
+        };
+        bitmap.first_available_id = (0..len)
+            .find(|&i| !bitmap.bit_slice()[i as usize])
+            .unwrap_or(len);
+        bitmap
+    }
+
+    /// Returns the length of the ID bitmap, i.e., the maximum number of IDs.
+    #[expect(unused)]
+    pub const fn len(&self) -> u16 {
+        self.len
+    }
+
+    /// Returns the maximum length of an ID bitmap.
+    pub const fn max_len() -> u16 {
+        BLOCK_SIZE as u16 * 8
+    }
+
+    /// Returns a reference to the underlying buffer of `BLOCK_SIZE` bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.buf
+    }
+
+    fn bit_slice(&self) -> &BitSlice<u8, Lsb0> {
+        &self.buf.as_bits()[..self.len as usize]
+    }
+
+    fn bit_slice_mut(&mut self) -> &mut BitSlice<u8, Lsb0> {
+        &mut self.buf.as_mut_bits()[..self.len as usize]
+    }
+
+    /// Returns true if the `id` is allocated.
+    ///
+    /// # Panics
+    ///
+    /// If the `id` is out of bounds, this method will panic.
+    pub fn is_allocated(&self, id: u16) -> bool {
+        self.bit_slice()[id as usize]
+    }
+
+    /// Allocates and returns a new `id`.
+    ///
+    /// If allocation is not possible, it returns `None`.
+    pub fn alloc(&mut self) -> Option<u16> {
+        if self.first_available_id < self.len {
+            let id = self.first_available_id;
+            self.bit_slice_mut().set(id as usize, true);
+            self.first_available_id = (id + 1..self.len)
+                .find(|&i| !self.bit_slice()[i as usize])
+                .unwrap_or(self.len);
+            Some(id)
+        } else {
+            None
+        }
+    }
+
+    /// Allocates a consecutive range of new `id`s.
+    ///
+    /// The `count` is the number of consecutive `id`s to allocate. If it is 0, return `None`.
+    ///
+    /// If allocation is not possible, it returns `None`.
+    pub fn alloc_consecutive(&mut self, count: u16) -> Option<Range<u16>> {
+        if count == 0 {
+            return None;
+        }
+
+        // Scan the bitmap from the position `first_available_id`
+        // for the first `count` number of consecutive 0's.
+        let allocated_range = {
+            // Invariance: all bits within `curr_range` are 0's
+            let mut curr_range = self.first_available_id..self.first_available_id + 1;
+            while curr_range.len() < count as usize && curr_range.end < self.len {
+                if !self.is_allocated(curr_range.end) {
+                    curr_range.end += 1;
+                } else {
+                    curr_range = curr_range.end + 1..curr_range.end + 1;
+                }
+            }
+
+            if curr_range.len() < count as usize {
+                return None;
+            }
+
+            curr_range
+        };
+
+        // Set every bit to 1 within the allocated range
+        for id in allocated_range.clone() {
+            self.bit_slice_mut().set(id as usize, true);
```

Comment:
Do not call `bit_slice_mut` in a loop. There could be potentially an overhead associated with the conversion. Review all methods for this flaw.